### PR TITLE
Expose service latency info in config file

### PR DESCRIPTION
**Pull request by variadico**
Currently, the config file doesn't recognize the `latency` config block in account exports. This change exposes those settings in the config file.

TODO:
- [x] return a `*serviceLatency`
- [x] rename `export.latency` to `export.lat`
- [x] delete `AddServiceExportWithResponseAndLatency`
- [x] fix lints
- [x] make sampling default `100`
- [x] allow `latency` to be string
- [x] accept both `int(100)` and `string(100%)`
- [x] test for explicitly setting sampling and letting the default fall through
- [x] use `TrackServiceExportWithSampling` to add latency to `exportAuth`

/cc @nats-io/core

### CODE REVIEW

File: server/accounts.go [hunk 1]
```diff
@@ -416,6 +417,12 @@ func (a *Account) AddServiceExportWithResponse(subject string, respType ServiceR
 	return nil
 }
 
+// AddServiceExportWithresponse will configure the account with the defined
+// export and response type.
+func (a *Account) AddServiceExportWithResponse(subject string, respType ServiceRespType, accounts []*Account) error {
+	return a.AddServiceExportWithResponseAndLatency(subject, respType, serviceLatency{}, accounts)
+}
+
```
Comment (variadico) [inline server/accounts.go]: Not too sure about this part. I created another function so I could pass `serviceLatency` to `exportAuth`. Is this okay? Or should I just add a param to the original function? Seems like we would break the API if we add a param, but not sure if we allow people to use those functions on their own.

File: server/configs/accounts.conf [hunk 2]
```diff
@@ -41,8 +41,15 @@ accounts: {
...
       {service: "nats.sub"}
+      { service: "nats.add"
+        response: singleton
+        latency: {
+          sample: 100%
+          subject: "latency.tracking.add"
+        }
+        accounts: [cncf]
+      }
     ]
```
Comment (variadico) [inline server/configs/accounts.conf]: I wasn't 100% sure if I should have added this here or if I should have created `latency.conf` or maybe some other way.

File: server/accounts.go [hunk 3]
```diff
@@ -376,8 +376,9 @@ func (a *Account) AddServiceExport(subject string, accounts []*Account) error {
 	return a.AddServiceExportWithResponse(subject, Singleton, accounts)
 }
 
-// AddServiceExportWithresponse will configure the account with the defined export and response type.
-func (a *Account) AddServiceExportWithResponse(subject string, respType ServiceRespType, accounts []*Account) error {
+// AddServiceExportWithresponse will configure the account with the defined
+// export, response type, and latency info.
+func (a *Account) AddServiceExportWithResponseAndLatency(subject string, respType ServiceRespType, latency serviceLatency, accounts []*Account) error {
```
Comment (derekcollison) [inline server/accounts.go]: There is a function to add on latency after you create the export. I would probably do that versus adding on to the above.

File: server/accounts.go (hunk 1 above)
Comment (derekcollison) [reply server/accounts.go]: There is a separate function that can add latency tracking to an export. Let's use that. You can find it in the original tests.

File: server/accounts_test.go [hunk 4]
```diff
@@ -452,6 +452,12 @@ func TestAccountParseConfigImportsExports(t *testing.T) {
 	if ea.respType != Singleton {
 		t.Fatalf("Expected to get a Singleton response type, got %q", ea.respType)
 	}
+	if want := "latency.tracking.add"; ea.latency.subject != want {
+		t.Fatalf("Expected to get latency subject %q, got %q", want, ea.latency.subject)
+	}
+	if want := int8(100); ea.latency.sampling != want {
```
Comment (derekcollison) [inline server/accounts_test.go]: I think 100 is default, let's make sure we test that and setting it specifically.

File: server/opts.go [hunk 5]
```diff
@@ -1246,10 +1246,11 @@ func setClusterPermissions(opts *ClusterOpts, perms *Permissions) {
...
 type export struct {
-	acc  *Account
-	sub  string
-	accs []string
-	rt   ServiceRespType
+	acc     *Account
+	sub     string
+	accs    []string
+	rt      ServiceRespType
+	latency serviceLatency
```
Comment (derekcollison) [inline server/opts.go]: Let's make it a * since most of the time it may not be present. Also short names so maybe lat.

File: server/opts.go [hunk 6]
```diff
@@ -1678,11 +1691,54 @@ func parseExportStreamOrService(v interface{}, errors, warnings *[]error) (*expo
...
 			}
 		}
-
 	}
 	return curStream, curService, nil
 }
 
+// parseServiceLatency returns a latency config block.
+func parseServiceLatency(v interface{}) (serviceLatency, error) {
+	latency, ok := v.(map[string]interface{})
```
Comment (derekcollison) [inline server/opts.go]: Maybe support simple version too. 

```
latency: <results.service.latency>
```

Which will default to 100 for sampling.

File: server/opts.go [hunk 7]
```diff
@@ -1678,11 +1691,54 @@ func parseExportStreamOrService(v interface{}, errors, warnings *[]error) (*expo
...
+
+	var sl serviceLatency
+
+	if v, ok := latency["sample"]; ok {
+		tok, ok := v.(token)
+		if !ok {
+			return serviceLatency{}, fmt.Errorf("Expected latency token, got %T", tok)
+		}
+
+		sampleStr, ok := tok.Value().(string)
```
Comment (derekcollison) [inline server/opts.go]: This should already be parsed as a number, not a string.

Comment (derekcollison) [reply server/opts.go]: We can support "100%" as well, but if I just do `sample: 50` we should parse that correctly.

File: server/opts.go [hunk 8]
```diff
@@ -1678,11 +1691,54 @@ func parseExportStreamOrService(v interface{}, errors, warnings *[]error) (*expo
...
+		sampleStr = strings.TrimSuffix(sampleStr, "%")
+		sample, err := strconv.ParseInt(sampleStr, 10, 8)
+		if err != nil {
+			return serviceLatency{}, fmt.Errorf("Failed to parse latency sample: %v", err)
+		}
+		sl.sampling = int8(sample)
+	}
+
+	if v, ok := latency["subject"]; ok {
+		tok, ok := v.(token)
```
Comment (variadico) [inline server/opts.go]: @derekcollison, does `subject` have a default value? Maybe this should throw an error if this is missing.

Comment (derekcollison) [reply server/opts.go]: No default for subject.

File: server/opts.go (hunk 7 above)
Comment (variadico) [reply server/opts.go]: Just so I understand, do we want to support
```
latency: {
  sample: 100   # <--- int
  subject: "latency.tracking.add"
}
```
and
```
latency: {
  sample: 100%  # <--- string
  subject: "latency.tracking.add"
}
```
¿?
Or just the int form?

Comment (derekcollison) [reply server/opts.go]: I like to be flexible in the config. I also think we support just `latency: "latency.metrics"`.

I think sample -> sampling.

File: server/opts.go (hunk 6 above)
Comment (variadico) [reply server/opts.go]: Ah, sorry. What does this mean? Is this an alternative config format than
```
    exports = [
      { service: "nats.add"
        response: singleton
        latency: {
          sample: 100%
          subject: "latency.tracking.add"
        }
        accounts: [cncf]
      }
    ]
```
¿?

Comment (variadico) [reply server/opts.go]: Ooh, based on your other comment, I think the answer is yes. Like this:
```
    exports = [
      { service: "nats.add"
        response: singleton
        latency: "latency.tracking.add"
        accounts: [cncf]
      }
    ]
```

Comment (derekcollison) [reply server/opts.go]: Yes correct.

File: server/const.go [hunk 9]
```diff
@@ -175,4 +175,8 @@ const (
 	// DEFAULT_ALLOW_RESPONSE_EXPIRATION is the default time allowed for a given
 	// dynamic response permission.
 	DEFAULT_ALLOW_RESPONSE_EXPIRATION = 2 * time.Minute
+
+	// DEFAULT_SERVICE_LATENCY_SAMPLE is the default sampling rate for service
+	// latency metrics
+	DEFAULT_SERVICE_LATENCY_SAMPLE = 100
```
Comment (variadico) [inline server/const.go]: Not sure if this variable has to be named this way. It's not the standard way consts are named in Go. :/

Comment (derekcollison) [reply server/const.go]: Yes I understand, just consistent with how we do it. We started when Go was very young ;)

File: server/accounts.go (hunk 1 above)
Comment (variadico) [reply server/accounts.go]: Okay, checked out some of the tests. Did you mean I could use `TrackServiceExportWithSampling`?

I tried something like this in the `parseAccounts` function.
```go
if err := service.acc.AddServiceExportWithResponse(service.sub, service.rt, accounts); err != nil {
	msg := fmt.Sprintf("Error adding service export %q: %v", service.sub, err)
	*errors = append(*errors, &configErr{tk, msg})
	continue
}

if service.lat != nil {
	err := service.acc.TrackServiceExportWithSampling(
		service.sub, service.lat.subject, int(service.lat.sampling),
	)
	if err != nil {
		msg := fmt.Sprintf("Error adding latency metrics for %s: %v", service.sub, err)
		*errors = append(*errors, &configErr{tk, msg})
		continue
	}
}
```

But, I keep hitting the error case.
```
Error adding latency metrics for nats.add: system account not setup
```

I noticed in the tests we use this.
```
if err := s.SetSystemAccount(globalAccountName); err != nil {
	t.Fatalf("Error setting system account: %v", err)
}
```

But, since I'm using the config file, am I going to need something like this, instead?
```
operator: /Users/synadia/.nsc/nats/memory/memory.jwt
resolver: MEMORY
resolver_preload: {
	ACSU3Q6LTLBVLGAQUONAGXJHVNWGSKKAUA7IY5TB4Z7PLEKSR5O6JTGR: eyJ0eXAiOiJqd3QiLCJhbGciOiJlZDI1NTE5In0.eyJqdGkiOiJPRFhJSVI2Wlg1Q1AzMlFJTFczWFBENEtTSDYzUFNNSEZHUkpaT05DR1RLVVBISlRLQ0JBIiwiaWF0IjoxNTU2NjU1Njk0LCJpc3MiOiJPRFdaSjJLQVBGNzZXT1dNUENKRjZCWTRRSVBMVFVJWTRKSUJMVTRLM1lERzNHSElXQlZXQkhVWiIsIm5hbWUiOiJBIiwic3ViIjoiQUNTVTNRNkxUTEJWTEdBUVVPTkFHWEpIVk5XR1NLS0FVQTdJWTVUQjRaN1BMRUtTUjVPNkpUR1IiLCJ0eXBlIjoiYWNjb3VudCIsIm5hdHMiOnsibGltaXRzIjp7InN1YnMiOi0xLCJjb25uIjotMSwibGVhZiI6LTEsImltcG9ydHMiOi0xLCJleHBvcnRzIjotMSwiZGF0YSI6LTEsInBheWxvYWQiOi0xLCJ3aWxkY2FyZHMiOnRydWV9fX0._WW5C1triCh8a4jhyBxEZZP8RJ17pINS8qLzz-01o6zbz1uZfTOJGvwSTS6Yv2_849B9iUXSd-8kp1iMXHdoBA
}

accounts: { ... }
```
¿?

Comment (derekcollison) [reply server/accounts.go]: You do not need full on operator mode, but yes you need a system account.

File: server/configs/accounts.conf (hunk 2 above)
Comment (derekcollison) [reply server/configs/accounts.conf]: Create a new one.

Comment (derekcollison) [reply server/configs/accounts.conf]: Or lately we just create them on the fly and delete them.

File: server/accounts.go [hunk 10]
```diff
@@ -508,10 +508,6 @@ func (a *Account) TrackServiceExportWithSampling(service, results string, sampli
 		return ErrBadPublishSubject
 	}
 
-	if a.srv == nil || !a.srv.eventsEnabled() {
-		return ErrNoSysAccount
-	}
```
Comment (wallyqs) [inline server/accounts.go]: This check had to be moved to be done earlier when processing the config at: https://github.com/nats-io/nats-server/pull/1130/files#diff-7b256a4f472c7d4b9e1931f2f1073b66R401-R409

File: server/opts.go [hunk 11]
```diff
@@ -398,6 +398,15 @@ func (o *Options) ProcessConfigFile(configFile string) error {
 	errors := make([]error, 0)
 	warnings := make([]error, 0)
 
+	// First check whether a system account has been defined, as that
+	// is a condition for other features to be enabled.
+	var hasSystemAccount bool
```
Comment (wallyqs) [inline server/opts.go]: We now do a quick check on whether the system account has been enabled, since rest of config map items will be processed randomly

File: server/config_check_test.go [hunk 12]
```diff
@@ -1113,6 +1113,72 @@ func TestConfigCheck(t *testing.T) {
...
+                    users = [
+                      { user : bar, pass: "" }
+                    ]
+
+                    exports = [
+                      { service: "nats.add" 
+                        response: singleton # chunked
+                        latency: {
+                          sampling: 100%
+                          subject: "latency.tracking.add"
```
Comment (wallyqs) [inline server/config_check_test.go]: when we try to use a wildcard subject here like `latency.results.*` we get an error like: 

```
nats-server: examples/latency-example.conf:4:1: Error adding service latency sampling for "nats.add": invalid publish subject
```

I think we should mention the subject to clarify that the bad subject is the one used for the latency samples

Comment (derekcollison) [reply server/config_check_test.go]: Yes it needs to be a valid publish subject, meaning literal.

File: server/accounts_test.go [hunk 13]
```diff
@@ -1305,12 +1305,6 @@ func TestAccountRequestReplyTrackLatency(t *testing.T) {
 	}
 
 	// Now let's add in tracking
-
```
Comment (derekcollison) [inline server/accounts_test.go]: Why did you remove this test?

File: server/config_check_test.go [hunk 14]
```diff
@@ -1113,6 +1113,72 @@ func TestConfigCheck(t *testing.T) {
 			errorLine: 0,
 			errorPos:  0,
 		},
+		{
+			name: "when setting latency tracking without a system account",
+			config: `
+                # system_account: sys
```
Comment (derekcollison) [inline server/config_check_test.go]: Commented out?

File: server/const.go (hunk 9 above)
Comment (derekcollison) [inline server/const.go]: SAMPLE -> SAMPLING

File: server/accounts_test.go (hunk 13 above)
Comment (variadico) [reply server/accounts_test.go]: @wallyqs

Comment (wallyqs) [reply server/accounts_test.go]: moved the check on whether system account is defined to be done earlier when processing the config at: https://github.com/nats-io/nats-server/pull/1130/files#diff-7b256a4f472c7d4b9e1931f2f1073b66R1462 but will try to add it back

Comment (derekcollison) [reply server/accounts_test.go]: I fixed this in JWT branch. Inside of `TrackServiceExportWithSampling` do this..

```go
	if a.srv != nil && !a.srv.eventsEnabled() {
		return ErrNoSysAccount
	}
```

Instead of what is there. On init the a.srv will be nil so skips the test.

Comment (variadico) [reply server/accounts_test.go]: Yeah, this seems to work!

Comment (derekcollison) [reply server/accounts_test.go]: Perfect.

File: server/config_check_test.go [hunk 15]
```diff
@@ -1113,6 +1113,70 @@ func TestConfigCheck(t *testing.T) {
 			errorLine: 0,
 			errorPos:  0,
 		},
+		{
+			name: "when setting latency tracking without a system account",
+			config: `
+                accounts {
+                  sys {
```
Comment (derekcollison) [inline server/config_check_test.go]: Maybe inline with one account per line.

`sys { users = [] }`

File: server/opts.go [hunk 16]
```diff
@@ -398,6 +398,15 @@ func (o *Options) ProcessConfigFile(configFile string) error {
...
+	if _, ok := m["system_account"]; ok {
```
Comment (derekcollison) [inline server/opts.go]: Maybe make a small utility function?

File: server/opts.go [hunk 17]
```diff
@@ -427,7 +436,7 @@ func (o *Options) ProcessConfigFile(configFile string) error {
 		case "disable_sublist_cache", "no_sublist_cache":
 			o.NoSublistCache = v.(bool)
 		case "accounts":
-			err := parseAccounts(tk, o, &errors, &warnings)
+			err := parseAccounts(tk, o, hasSystemAccount, &errors, &warnings)
```
Comment (derekcollison) [inline server/opts.go]: if we pass o and o is top level options, we can discover hasSystemAccount status inside yes?

File: server/opts.go [hunk 18]
```diff
@@ -1678,11 +1720,86 @@ func parseExportStreamOrService(v interface{}, errors, warnings *[]error) (*expo
...
+	// Read sampling value.
+	if v, ok := latency["sampling"]; ok {
+		tok, ok := v.(token)
+		if !ok {
+			return nil, &configErr{token: root,
+				reason: fmt.Sprintf("Expected latency token, got %T", tok)}
+		}
+
+		var sample int64
+		switch v := tok.Value().(type) {
```
Comment (derekcollison) [inline server/opts.go]: Is v set to type here va value?

File: server/opts.go (hunk 17 above)
Comment (wallyqs) [reply server/opts.go]: Right, if besides checking whether has system account we also set it at the beginning, would be possible to use the field from options, will update.

File: server/opts.go (hunk 18 above)
Comment (wallyqs) [reply server/opts.go]: I cleaned up some of the config and handling tokens here to report line correctly, should be fixed now.

File: server/config_check_test.go [hunk 19]
```diff
@@ -1113,6 +1113,105 @@ func TestConfigCheck(t *testing.T) {
 			errorLine: 0,
 			errorPos:  0,
 		},
+		{
+			name: "when setting latency tracking without a system account",
+			config: `
+                accounts {
+                  sys { users = [ {user: sys, pass: "" } ] }
+
+                  nats.io: {
```
Comment (derekcollison) [inline server/config_check_test.go]: Would place on same line as sys above.

File: server/config_check_test.go [hunk 20]
```diff
@@ -1113,6 +1113,105 @@ func TestConfigCheck(t *testing.T) {
...
+                    users = [ { user : bar, pass: "" } ]
+
+                    exports = [
+                      { service: "nats.add"
+                        response: singleton # chunked
```
Comment (derekcollison) [inline server/config_check_test.go]: Remove comment.

File: server/config_check_test.go [hunk 21]
```diff
@@ -1113,6 +1113,105 @@ func TestConfigCheck(t *testing.T) {
...
+			errorPos:  0,
+		},
+		{
+			name: "when setting latency tracking with an invalid publish subject",
+			config: `
+                system_account = sys
+                accounts {
+                  sys { users = [ {user: sys, pass: "" } ] }
+
+                  nats.io: {
```
Comment (derekcollison) [inline server/config_check_test.go]: Same here with one line.
